### ISSUE TITLE
Install missing package on the fly (for texlive)

### ISSUE THREAD

**Issue by vanabel** [lervag/vimtex#561]
We know that miktex has the feature install missing packages automatically, but it seems that texlive is not support this fancy feature (see, e.g. [here](http://tex.stackexchange.com/a/110533/19832)). Then I come here to ask, is there any possibility to include this as a new feature in vimtex?

**Comment by lervag** [comment]
True, this is not a feature of texlive. However, I don't think this should be a feature of vimtex either, because:

1. It will be very difficult to make it work on both Windows, OSX and Linux.
2. It does not _feel_ right. Vimtex is about editing LaTeX documents. Most features are directed towards making the editing session better. Further, it is important to keep things as simple as possible. Thus:
3. If anyone ever implemented such a feature, they could easily package it and share it as a dedicated plugin that would interact well with both vimtex, LaTeX-Box, latex-suite, etc. I think that would be a better solution. Although, I think even better:
4. Can't you just install texlive-full or similar, which includes most/all packages?

**Comment by dbmrq** [comment]
I have this in my `ftplugin/tex.vim`:

```vim
function! InstallPackages()
    let winview = winsaveview()
    call inputsave()
    let cmd = ['sudo -S tlmgr install']
    %call add(cmd, matchstr(getline('.'), '\\usepackage\(\[.*\]\)\?{\zs.*\ze\}'))
    echomsg join(cmd)
    let pass = inputsecret('Enter sudo password:') . "\n"
    echo system(join(cmd), pass)
    call inputrestore()
    call winrestview(winview)
endfunction

command! InstallPackages call InstallPackages()
```

So I can run `:InstallPackages` and it, well, installs the packages.

(I got it form my question [here](http://vi.stackexchange.com/q/8138/7642).)

**Comment by nasenatmer** [comment]
There are, however, two downsides to this method, dbmrq:

1. it requires tlmgr (in which package is it included? I don't have that on my Arch Linux system)

2. it installs packages via a different system than most Linux distributions do with their packaging system 
which are updated automatically by the package manager (apt-get, pacman, etc…)

**Comment by benutzer193** [comment]
tlmgr is part of the minimal install of texlive, so I guess you should have it installed.

**Comment by nasenatmer** [comment]
My Latex installation includes the following packages, a package `texlive-minimal` does not exist in Arch. 

```
texlive-core
texlive-bin
texlive-genericextra
texlive-humanities
texlive-bibtexextra
texlive-latexextra
```

Is `tlmgr` maybe the same as [texlive-localmanager](https://aur.archlinux.org/packages/texlive-localmanager-git/)?

**Comment by benutzer193** [comment]
No, thats another helper script, but should work as well, if you change the commands.

tlmgr should be in [texlive-core](https://www.archlinux.org/packages/extra/any/texlive-core/) if you check the filelist.

**Comment by nasenatmer** [comment]
> tlmgr should be in texlive-core if you check the file list.

Hah, turns out my installation of texlive-core was incomplete. I reinstalled it and now I have it. Thanks for your hint and sorry for the noise!

**Comment by lervag** [comment]
Just to make it clear: I think this discussion is interesting, but I still don't think this should be added to vimtex. However, I can see that this is something that some people might find useful, and so this discussion might be useful to them as a reference.

**Comment by vanabel** [comment]
@dbmrq great, but in there are exceptional cases, in which the package name and the style name is not the same as package name or the package is load by another package. for example, I never used the `titletoc.sty` but I need it in compile my thesis (which provided a `cls` file, in which there are a lot of package loaded), while I try to install it by `tlmgr install titletoc`, it fails to find any package, since the right package is `titlesec`.

**Comment by dbmrq** [comment]
@vanabel Yes, I know it's not perfect and I'd also love a better option, but in the meanwhile I thought this could help. Let me know if you find a better way. :)

**Comment by gabrielbdsantos** [comment]
I know this is an old topic but I stumbled upon this issue a few days ago looking for an alternative to manually install every missing dependency. Well, I came back to share this very interesting package [`texliveonfly`](https://www.ctan.org/pkg/texliveonfly). I've just tested it in a bunch of projects I have here, and it resolves almost every missing dependency. It can handle even those case where `some-file.sty` is not a package _per se_ but a file in package `package-containing-some-file`.
